src/librbd/cache/rwl/ShutdownRequest.cc

```diff
@@ -0,0 +1,144 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "librbd/cache/rwl/ShutdownRequest.h"
+#include "librbd/Utils.h"
+#include "common/dout.h"
+#include "common/errno.h"
+#include "librbd/Operations.h"
+#include "librbd/asio/ContextWQ.h"
+#include "librbd/cache/ReplicatedWriteLog.h"
+
+
+#define dout_subsys ceph_subsys_rbd_rwl
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::cache::rwl:ShutdownRequest: " \
+                           << this << " " << __func__ << ": "
+
+namespace librbd {
+namespace cache {
+namespace rwl {
+
+using util::create_async_context_callback;
+using util::create_context_callback;
+
+template <typename I>
+ShutdownRequest<I>* ShutdownRequest<I>::create(I &image_ctx,
+                                       Context *on_finish) {
+  return new ShutdownRequest(image_ctx, on_finish);
+}
+
+template <typename I>
+ShutdownRequest<I>::ShutdownRequest(I &image_ctx, Context *on_finish)
+  : m_image_ctx(image_ctx),
+    m_on_finish(create_async_context_callback(image_ctx, on_finish)),
+    m_error_result(0) {
+}
+
+template <typename I>
+ShutdownRequest<I>::~ShutdownRequest() {
+}
+
+template <typename I>
+void ShutdownRequest<I>::send() {
+  send_shutdown_image_cache();
+}
+
+template <typename I>
+void ShutdownRequest<I>::send_shutdown_image_cache() {
+  CephContext *cct = m_image_ctx.cct;
+  ldout(cct, 10) << dendl;
+
+  if (m_image_ctx.image_cache == nullptr) {
+    finish();
+    return;
+  }
+
+  using klass = ShutdownRequest<I>;
+  Context *ctx = create_context_callback<klass, &klass::handle_shutdown_image_cache>(
+    this);
+
+  m_image_ctx.image_cache->shut_down(ctx);
+}
+
+template <typename I>
+void ShutdownRequest<I>::handle_shutdown_image_cache(int r) {
+  CephContext *cct = m_image_ctx.cct;
+  ldout(cct, 10) << dendl;
+
+  if (r < 0) {
+    save_result(r);
+  } else {
+    delete m_image_ctx.image_cache;
+    m_image_ctx.image_cache = nullptr;
+  }
+  send_remove_image_cache_state();
+}
+
+template <typename I>
+void ShutdownRequest<I>::send_remove_image_cache_state() {
+  CephContext *cct = m_image_ctx.cct;
+  ldout(cct, 10) << dendl;
+
+  using klass = ShutdownRequest<I>;
+  Context *ctx = create_context_callback<klass, &klass::handle_remove_image_cache_state>(
+    this);
+  std::shared_lock owner_lock{m_image_ctx.owner_lock};
+  m_image_ctx.operations->execute_metadata_remove(".librbd/image_cache_state", ctx);
+}
+
+template <typename I>
+void ShutdownRequest<I>::handle_remove_image_cache_state(int r) {
+  CephContext *cct = m_image_ctx.cct;
+  ldout(cct, 10) << dendl;
+
+  if (r < 0) {
+    save_result(r);
+  } else {
+    m_image_ctx.cache_state = nullptr;
+  }
+  send_remove_feature_bit();
+}
+
+template <typename I>
+void ShutdownRequest<I>::send_remove_feature_bit() {
```

Comment:
Nit: should you remove the feature bit before removing the state so that a crash doesn't leave you w/ zero state but the feature bit enabled?

Reply:
During InitRequest<I>::pre_check(), if rwl_enabled in ceph config and image cache feature bit are false, it skips the handler of image cache.  So if we remove the feature bit before removing the cache state, it may lead the cache state never been removed later.  Unless we check whether cache state exists everytime feature bit is false, and remove if cache state exists. 
  
Currently in the code, if the config rwl_enabled is true and the feature bit is set, it exits with error. But we can enhance it later: when image cache state is null, unset feature bit.  
https://github.com/ceph/ceph/blob/1c2c6489b61b09ecdb1a41ab54b467e9973e0d74/src/librbd/cache/rwl/InitRequest.cc#L68-L74__

Reply:
Now that the feature bit is used to track a dirty cache that needs playback, why should `InitRequest::pre_check` skip initialization of the RWL cache if the feature bit isn't set?

Reply:
@dillaman Sorry, I am confused about your question.   
Currently, in the codes, it removes image cache state before removing feature bit. Please look at the following table. When the feature bit isn't set, it means the RWL cache is shut down gracefully. So it just skips the RWL's initialization if the rwl config is not enabled.

When the config rwl_enabled is not true: 
Feature bit | cache state | how to handle 
------------ | ------------- | -------------
1 | existed | Initialize rwl, flush data (TBD)
1 | removed| Unset feature bit  when restarting librbd. (TBD). This happens when librbd crashes right after removing the cache state. 
0 | exists| The case will never happen
0 | removed| skip initialization of the RWL

But if it removes the feature bit before removing the cache state as you suggest. The above case 3 may happen. It may lead that the cache state metadata never be deleted. 
Feature bit | cache state | how to handle 
------------ | ------------- | -------------
0 | exists|

Reply:
> Currently, in the codes, it removes image cache state before removing feature bit. Please look at the following table. When the feature bit isn't set, it means the RWL cache is shut down gracefully. So it just skips the RWL's initialization if the rwl config is not enabled.

I am not questioning how it's currently implemented. I am asking if it's currently implemented correctly. If you crash after removing the image-meta state but before removing the feature bit, what does that mean for the next client that tries to open the image? How will this feature bit get cleared?

Reply:
You are right. I would like to remove feature the bit before removing the image-meta state. This is easy to handle, and also reasonable between different clients.   
When reopening the image cache, if the feature bit is unset but the image-meta state exists, the image-meta will be ignored.